src/engine/Operation.h

```diff
@@ -316,8 +316,11 @@ class Operation {
 
   const auto& getLimit() const { return _limit; }
 
+  // Actual implementation of `clone()` without extra checks.
+  virtual std::unique_ptr<Operation> cloneImpl() const = 0;
+
```

Comment:
Should be private (at least in the base class, but preferably also in all the child classes)